homeassistant/components/kodi/media_player.py

```diff
@@ -227,6 +228,7 @@ async def async_setup_entry(
 
     entity = KodiEntity(connection, kodi, name, uid)
     async_add_entities([entity])
+    async_add_entities(entity.sensor_entities)
```

Comment:
The binary sensor entities should be added `async_setup_entry` for the relevant platform. This will require some refactoring

Reply:
Hello @OnFreund and already thank you for taking the time to have a look. As you can see this is my first contribution and I have to admit that the closer I get to the core the less I do understand at the moment.

Let me outline what I understood from you comment and some questions in it:

- [ ] create the "relevant platform" which would be a file: binary_sensor.py
- [ ] import stuff
- [ ] make the async setup:
```
async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
```
  * This is called automatically by the core?
  * Below would be the same data as in the media_player.py?
```
    data = hass.data[DOMAIN][config_entry.entry_id]
    connection = data[DATA_CONNECTION]
    kodi = data[DATA_KODI]
    name = config_entry.data[CONF_NAME]
    if (uid := config_entry.unique_id) is None:
        uid = config_entry.entry_id
```

- [ ] Create
```
class KodiBinarySensor()
```
- [ ] Get notified when it is added to HA
```
 async def async_added_to_hass(self) -> None:
        """Connect the websocket if needed."""
        if not self._connection.can_subscribe:
            return

        if self._connection.connected:
            await self._on_ws_connected()
[...]
```
  * Here I'm a little lost about how to handle the WebSocket connection stuff. Which is all managed in the media_player.py file. That is the main reason, why I piggy backed my code there.

If I'm correct with all this until now, what is the minimum I need to do to have:
* always the right ws connection?
* the callbacks register on every connect/reconnect/disconnect scenario?

Reply:
Hi @OnFreund , I did the refactoring you asked for, and to my surprise everything seams to work.

For the ws connection I started with the most simple approach and just added the callbacks to the connection without checking the connection status at all. 

## Tests

Acceptance criteria are always: 
* wait for the screensaver to active -> check status in HA is on
* deactivate the screensaver by user input -> check status in HA is off

- [x] HA running then restart Kodi
- [x] HA restart while Kodi is running
- [x] HA restart while Kodi is stopped; wait for HA is running; start Kodi

## Possible improvements for another PR
* Trying to get the current state (e.g. screensaver) after initial connection. Some findings but seams more complicated:
  https://community.openhab.org/t/kodi-addon-how-to-detect-screensaver-idle-time/109259
  https://kodi.wiki/view/List_of_boolean_conditions